File: 2do Curso/MP - Metodologia de la Programacion/Practica1/src/Cliente.cpp
#include "Cliente.h"

Cliente::Cliente( long int dni, char* nombre, const Fecha& f):f(f)
{
    this->dni=dni;

    this->nombre=new char[strlen(nombre)+1];
    strcpy(this->nombre,nombre);
    //ctor
}

Cliente::Cliente(const Cliente& c):f(c.f){

this->dni=c.dni;
this->nombre=new char[strlen(c.nombre)+1];
strcpy(this->nombre,c.nombre);

}

void Cliente::setNombre(const char* nombre){

if(this->nombre!=NULL)
    delete [] this->nombre;

this->nombre=new char[strlen(nombre)+1];
strcpy(this->nombre,nombre);

}

void Cliente::setFecha(const Fecha& f){

this->f.setFecha(f.getDia(),f.getMes(),f.getAnio());

}

Cliente& Cliente::operator=(const Cliente& c){

if(this!=&c){
    //si no es el mismo cliente hago la igualdad
    this->dni=c.dni;
    delete [] this->nombre;
    this->nombre=new char[strlen(c.nombre)+1];
    strcpy(this->nombre,c.nombre);
    this->f=c.f;
}

return *this;

}


bool Cliente::operator==(const Cliente& c) const{

if(this->dni!=c.dni)
    return false;
if(strcmp(this->nombre,c.nombre)!=0)
    return false;
if(this->f.getDia()!=c.f.getDia() || this->f.getMes()!=c.f.getMes() || this->f.getAnio()!=c.f.getAnio())
    return false;
return true;

}

ostream& operator <<(ostream& s,const Cliente& c){

s<< c.getNombre() << " (" << c.getDni() << " - " << c.getFecha() << ")" << endl;

return s;
}

Cliente::~Cliente()
{
    delete [] this->nombre;
    //dtor
}

File: 2do Curso/MP - Metodologia de la Programacion/Practica2/src/libClases/Cliente.java
package libClases;

public class Cliente implements Cloneable,Proceso {
	private static int contador=1;
	private final String nif;
	private final int codCliente;
	private String nombre;
	private final Fecha fechaNac;
	private final Fecha fechaAlta;
	
	private static final Fecha FechaDefecto= new Fecha(1,1,2018);
	
	//Constructores
	public Cliente(String NIF, String nom, Fecha fNac, Fecha fAlta)
	{
		this.nif=NIF;
		this.nombre=nom;
		this.fechaNac=(Fecha)fNac.Clone();
		this.fechaAlta=(Fecha)fAlta.Clone();
		this.codCliente=contador++;
	}
	
	public Cliente (String NIF, String nom, Fecha fNac) {
		
		Fecha fAlta= new Fecha(01,01,2018);
		this.nif=NIF;
		this.nombre=nom;
		this.fechaNac=(Fecha)fNac.Clone();
		this.fechaAlta=(Fecha)fAlta.Clone();
		this.codCliente=contador++;
	}
	
	public Cliente(Cliente c)
	{
		this.nif=c.nif;
		this.nombre=c.nombre;
		this.fechaNac=(Fecha)c.fechaNac.Clone();
		this.fechaAlta=(Fecha)c.fechaAlta.Clone();
		this.codCliente=contador++;
	}
	
	//Getters publicos de la clase
	public String getNIF() {return nif;};
	public int getCodCliente() {return codCliente;};
	public String getNombre() {return nombre;};
	public Fecha getFechaNac() {return (Fecha)fechaNac.Clone();};
	public Fecha getFechaAlta() {return (Fecha)fechaAlta.Clone();};
	public static Fecha getFechaPorDefecto() {return (Fecha)FechaDefecto.Clone();};
	
	
	//Setters publicos de la clase, afectan al nombre del cliente, al dni y a la fecha de alta
	public void SetFechaAlta(Fecha f) {
		this.fechaAlta.setFecha(f.getDia(),f.getMes(),f.getAnio());
	}
	
	public static void SetFechaDefecto(Fecha f) {
		FechaDefecto.setFecha(f.getDia(),f.getMes(),f.getAnio());
	}
	
	public void SetNombre(String nom)
	{
		this.nombre=nom;
	}
	
	public float factura() {return 0;}
	
	
	public Object Clone()
	{
		return new Cliente(this);
	}
	
	
	public void ver()
{
		System.out.println(this);
}
	
	public String toString() 
	{
		return nif + " " + fechaNac+": " + nombre + " (" + codCliente + " - " + fechaAlta + ") ";
	}
	
	public boolean equals(Object o) {
		return o instanceof Cliente && nif.equals(((Cliente)o).nif);
	}
}
